feat: GPKG merger for combining two gpkg's

roicohen326 opened the pull request:
cr

roicohen326 committed:
prepare code for review

cr

razbroc commented:
Hi Roi!
Something you should know- there are some git conventions that we follow in MapColonies.
Pr's title should be "feat: GPKG merger for combining two gpkg's".
This convention also applies to git commits.
These are the subjects you can use:

build: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
ci: Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
docs: Documentation only changes
feat: A new feature
fix: A bug fix
perf: A code change that improves performance
refactor: A code change that neither fixes a bug nor adds a feature
style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, lint, etc)
test: Adding missing tests or correcting existing tests

razbroc commented:
Make sure you fulfill the right column for reviewers, assignee, etc...

razbroc commented on .DS_Store:
can this also be included in .gitignore?

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
-const file2 = './data/test_south.gpkg';
-const output = './merged_complete.gpkg';
+const file1 = './data/Syria.gpkg';
+const file2 = './data/blueMarble.gpkg';
```
dynamic

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
-const output = './merged_complete.gpkg';
+const file1 = './data/Syria.gpkg';
+const file2 = './data/blueMarble.gpkg';
+const output = './merged_syria_bluemarble.gpkg';
```
make this output name using file1 and file2 names. dynamically.

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
-  if (!fs.existsSync(file2)) throw new Error('File 2 missing');
-  console.log('✅ Files validated');
+  console.log('Step 1: Validate new GeoPackage files...');
+  if (!fs.existsSync(file1)) throw new Error('Syria file missing');
```
this is a great check! 
Check out our error handling libraries - [mapcoloneis](https://github.com/MapColonies/error-express-handler/blob/master/src/index.ts)

Which error code would you use in this case?

roicohen326 replied:
I used StatusCodes.BAD_REQUEST (400) because file path validation failures represent client-side input errors. Since the user provided invalid file paths as arguments, this falls under the 4xx category for client errors.

razbroc replied:
nice

razbroc commented on merge.ts:
I noticed many hard-coded things in this file that should be inputable and more dynamic.

I will be writing 'dynamic' on them- I reference making them changeable in case I have other files to merge.

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
-const file1 = './data/test_north.gpkg';
-const file2 = './data/test_south.gpkg';
-const output = './merged_complete.gpkg';
+const file1 = './data/Syria.gpkg';
```
dynamic

shimoncohen replied:
Don't write these kind of things in-code, even if it's not meant to be configurable.

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
-  if (!fs.existsSync(file2)) throw new Error('File 2 missing');
-  console.log('✅ Files validated');
+  console.log('Step 1: Validate new GeoPackage files...');
+  if (!fs.existsSync(file1)) throw new Error('Syria file missing');
```
dynamic error- for example:

``${file1} file missing``

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
-  console.log('✅ Files validated');
+  console.log('Step 1: Validate new GeoPackage files...');
+  if (!fs.existsSync(file1)) throw new Error('Syria file missing');
+  if (!fs.existsSync(file2)) throw new Error('Blue Marble file missing');
```
dynamic error

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+  if (!fs.existsSync(file1)) throw new Error('Syria file missing');
+  if (!fs.existsSync(file2)) throw new Error('Blue Marble file missing');
+  
+  const size1 = (fs.statSync(file1).size / 1024 / 1024).toFixed(2);
```
fix [magic number 1024?](https://en.wikipedia.org/wiki/Magic_number_(programming))

shimoncohen replied:
Add comments to lines like this, what are you trying to validate?

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+  if (!fs.existsSync(file2)) throw new Error('Blue Marble file missing');
+  
+  const size1 = (fs.statSync(file1).size / 1024 / 1024).toFixed(2);
+  const size2 = (fs.statSync(file2).size / 1024 / 1024).toFixed(2);
```
#https://github.com/roicohen326/geo-package-merger/pull/2/files#r2121732061

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+  
+  const size1 = (fs.statSync(file1).size / 1024 / 1024).toFixed(2);
+  const size2 = (fs.statSync(file2).size / 1024 / 1024).toFixed(2);
+  console.log(`✅ Syria dataset: ${size1} MB`);
```
dynamic for syria

razbroc replied:
dynamic - no emojis

shimoncohen replied:
Also add the word for the type of information, like `${file} dataset size: ${size}`.

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+  const size1 = (fs.statSync(file1).size / 1024 / 1024).toFixed(2);
+  const size2 = (fs.statSync(file2).size / 1024 / 1024).toFixed(2);
+  console.log(`✅ Syria dataset: ${size1} MB`);
+  console.log(`✅ Blue Marble dataset: ${size2} MB`);
```
dynamic bluemarble

razbroc replied:
dynamic - no emojis

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+  console.log(`✅ Syria dataset: ${size1} MB`);
+  console.log(`✅ Blue Marble dataset: ${size2} MB`);
 
   console.log('Step 2: Copy base file...');
```
it important to make logs as informative as possible- for example add the name of the file you are refrencing.

shimoncohen replied:
Yes, including source and destination.

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
   if (fs.existsSync(output)) fs.unlinkSync(output);
   fs.copyFileSync(file1, output);
-  console.log('✅ Base file copied');
+  console.log('✅ Base file copied (Syria as base)');
```
dynamic

razbroc replied:
dynamic - no emojis

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+  console.log(`✅ Blue Marble dataset: ${size2} MB`);
 
   console.log('Step 2: Copy base file...');
   if (fs.existsSync(output)) fs.unlinkSync(output);
```
if statements should **ALWAYS** have curly braces as it is best practice.

Also read about [ternary operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_operator)

razbroc replied:
Once you make naming dynamic, you won't need to remove old files.
Other than that, this always removes the previous merge, which in many cases we would probably want to avoid doing such an action on the data, as it might still be relevant for the user.

Please remove it and find another solution to keeping previous files.

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
-  console.log('✅ Base file copied');
+  console.log('✅ Base file copied (Syria as base)');
 
   console.log('Step 3: Open databases...');
```
dynamic

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
 
-  console.log('Step 4: Get table info...');
-  const tables = sourceDb.prepare(`
+  console.log('Step 4: Analyze both datasets...');
```
`Analyze` is meaningless to me

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+    AND name NOT LIKE 'sqlite_%' 
+    AND name NOT LIKE 'gpkg_%'
+    AND name NOT LIKE 'rtree_%'
+  `).all();
```
Both queries are identical.
read about [code duplication](https://linearb.io/blog/code-duplication) and fix it accordingly

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+  console.log('Step 4: Analyze both datasets...');
+  
+  // Get tables from both databases
+  const targetTables = targetDb.prepare(`
```
Not sure the use of prepare here is optimal.
Read about it some more and let me know as it is an object representing a single SQL statement.

roicohen326 replied:
I researched .prepare() in better-sqlite3 and understand your concern about potential overkill for single-use queries. Unlike other SQL libraries, better-sqlite3 requires .prepare() for all select queries that return data - there's no direct .all() method on the database object. Even for single-use queries, .prepare().all() is the mandatory pattern. The library is designed this way for consistency and performance.

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+    AND name NOT LIKE 'rtree_%'
+  `).all();
+  
+  console.log(`📊 Syria has ${targetTables.length} data tables`);
```
emojis in logs are not best practice. also make `Syria` dynamic.

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+  
+  console.log(`📊 Syria has ${targetTables.length} data tables`);
+  for (const table of targetTables) {
+    const count = targetDb.prepare(`SELECT COUNT(*) as count FROM ${(table as any).name}`).get() as any;
```
`table as any` is not good.
you assume it has `name` property.
if `tagetTables` query result will not have `name` propery (query returns something else- or empty) your program will crash as `name` does not exist.

make `table` property TYPED.
see [following page](https://www.typescriptlang.org/docs/handbook/2/objects.html)

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
   console.log('Step 3: Open databases...');
   const targetDb = new Database(output);
   const sourceDb = new Database(file2, { readonly: true });
   console.log('✅ Databases opened');
```
emojis

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+        const createTableSQL = sourceDb.prepare(`
+          SELECT sql FROM sqlite_master 
+          WHERE type='table' AND name = ?
+        `).get(tableName) as any;
```
Both queries are identical.
read about [code duplication](https://linearb.io/blog/code-duplication) and fix it accordingly

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+      
+      if (!tableExists) {
+        // Create table structure from source
+        console.log(`🆕 Creating new table: ${tableName}`);
```
emoji

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+        WHERE type='table' AND name = ?
+      `).get(tableName);
+      
+      if (!tableExists) {
```
If the table does not exist, we continue as usual?
What about errors or handling this matter?

roicohen326 replied:
Good point about error handling. The code now has comprehensive try/catch blocks around the entire table processing operation, checks for valid table creation SQL before executing, and logs any errors while continuing with other tables. If table creation fails for any reason, the error is caught and logged, ensuring the merge process continues with remaining tables.

razbroc replied:
Having a  try/catch block around the entire table processing operation is not best practice. It will be harder to isolate and make the logging and error handling specific for each different action harder

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+          WHERE type='table' AND name = ?
+        `).get(tableName) as any;
+        
+        if (createTableSQL?.sql) {
```
refrencing 
https://github.com/roicohen326/geo-package-merger/pull/2/files#r2123894859

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+  console.log(`📊 Syria has ${targetTables.length} data tables`);
+  for (const table of targetTables) {
+    const count = targetDb.prepare(`SELECT COUNT(*) as count FROM ${(table as any).name}`).get() as any;
+    console.log(`   - ${(table as any).name}: ${count.count} rows`);
```
same as above https://github.com/roicohen326/geo-package-merger/pull/2/files#r2123894859

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+        
+        if (createTableSQL?.sql) {
+          targetDb.exec(createTableSQL.sql);
+          console.log(`✅ Table structure created`);
```
emoji +  add info

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+      // Insert data (different datasets shouldn't have ID conflicts)
       const result = targetDb.prepare(`
-        INSERT OR IGNORE INTO ${tableName} 
+        INSERT INTO ${tableName} 
```
`tableName` is typed "any" and should be typed "string".

as for now, it might be null or object and crash the program.

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
       const result = targetDb.prepare(`
-        INSERT OR IGNORE INTO ${tableName} 
+        INSERT INTO ${tableName} 
         SELECT * FROM source_db.${tableName}
```
tableName is typed "any" and should be typed "string".

as for now, it might be null or object and crash the program.

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
       `).run();
       
-      console.log(`✅ Merged ${result.changes} rows from ${tableName}`);
+      console.log(`✅ Added ${result.changes} rows from ${tableName}`);
```
this is better, just remove emoji

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+  console.log('Step 5: Attach and merge different datasets...');
   targetDb.exec(`ATTACH DATABASE '${file2}' AS source_db`);
   
   let totalMerged = 0;
```
It's best practice to use `const` instead of `let`, as you modify it only once. I suggest figuring it out and making it `const`.

Read [HERE](https://stackoverflow.com/questions/41086633/in-javascript-why-should-i-usually-prefer-const-to-let) why const is better and important to use in code.

roicohen326 replied:
I understand the preference for const, but totalMerged cannot be const because it's being incremented multiple times in the loop (once per table processed).

razbroc replied:
Movan

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
       
     } catch (err) {
-      console.log(`⚠️ Skipped table: ${err}`);
+      console.log(`⚠️ Error with table ${(table as any).name}: ${err}`);
```
This is a general error handling. notice other comments to separate the process into micro processes to handle errors better and make code flow more readable and stand within [open close princple](https://www.freecodecamp.org/news/open-closed-principle-solid-architecture-concept-explained/)

roicohen326 replied:
Thank you , this is a great point about code architecture. I've refactored the table processing into smaller, focused functions following the Open/Closed Principle.

razbroc replied:
goood

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
   }
 
   targetDb.exec('DETACH DATABASE source_db');
   sourceDb.close();
```
You would also want to close db connection on errors that crash the program.

Also read about optimal handling db - when should you close and open the connection.
For example, a quick recap.

✅ Best Practice (General Rule)
Close the DB connection after each transaction only if:

You open a new connection per transaction.

You don’t need persistent connections (e.g., CLI scripts, short-lived jobs).

Keep the connection open and close it at the end of the program if:

You use a single, long-lived connection (e.g., web server, background worker).

You manage transactions within that connection.

You're using a connection pool.

💡 Recommended Approach
For most modern applications:

Use a connection pool.

Acquire a connection from the pool per transaction or request.

Release (not close) the connection after the transaction.

Let the pool handle actual closing at program shutdown.

🚫 Avoid:
Leaving connections open too long without activity (can cause timeouts or resource leaks).

Reopening and closing the DB connection repeatedly if it adds performance overhead.

If you tell me what kind of app you're building (web server, script, etc.), I can give a more specific recommendation. 

please fix accordingly.

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
   
-  console.log(`🎉 MERGE COMPLETE! Total rows merged: ${totalMerged}`);
-  console.log(`Output: ${output}`);
+  const finalSize = (fs.statSync(output).size / 1024 / 1024).toFixed(2);
```
https://github.com/roicohen326/geo-package-merger/pull/2/files#r2121732061

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
-  console.log(`🎉 MERGE COMPLETE! Total rows merged: ${totalMerged}`);
-  console.log(`Output: ${output}`);
+  const finalSize = (fs.statSync(output).size / 1024 / 1024).toFixed(2);
+  console.log(`\n🎉 SYRIA + BLUE MARBLE MERGE COMPLETE!`);
```
DYNAMIC + EMOJIS!

razbroc commented on merge.ts:
```diff
@@ -1,66 +1,115 @@
+  console.log(`📍 Output: ${output} (${finalSize} MB)`);
+  console.log(`🌍 Combined: Syria geographic data + Blue Marble imagery!`);
 
 } catch (error: any) {
```
separate process to functions, make try catch precise on failure points, and not for the whole program.
This is not best practice error handling.

razbroc commented:
Please add git workflows for lint checks

razbroc commented:
Well done Roi.
Remember - comments are the best way to improve.
Take each comment seriously as a place to learn and improve.

I'm satisfied as this task is a great opportunity to learn relevant everyday coding techniques that you will use every day.
The key things that I want you to learn from these errors are:
* TypeScript types and why it is important for every variable to be typed.
* Error handling and best practices
* Logging best practices
* Coding with databases best practices
* code duplications
* dynamic code - open close principle. 

Keep up the motivation!

razbroc reviewed (changes requested):
keep it up!

razbroc commented:
Please make `.Dockerfile` so it can be run as a docker image (allow configurable environment variable).

roicohen326 committed:
Update merge.ts

feat: make GeoPackage merger fully dynamic with TypeScript interfaces

- Replace hard-coded paths with CLI arguments
- Add type safety with interfaces
- Remove code duplication and magic numbers
- Improve error handling and logging

roicohen326 committed:
Update .gitignore

roicohen326 committed:
Update package.json

roicohen326 committed:
Update package.json

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+
+// Show usage if not enough arguments
+if (args.length < 2) {
+  console.log('\nUsage:');
```
we should avoide writing multiple logs right after the other.
use one console log.

please add a messege that will fit the log, ex "not enough arguments. try...."

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
-  const sourceDb = new Database(file2, { readonly: true });
-  console.log('✅ Databases opened');
+// Parse command line arguments
+const args = process.argv.slice(2);
```
Nice concept. Please add a readme that will explain how to add the arguments to run the program.

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+const file2 = args[1];
 
-  console.log('Step 4: Analyze both datasets...');
+// Create dynamic output filename based on input filenames
```
We should avoid having notes in the code.
A good code should be self-explanatory without adding words and explanations.

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
-  `).all();
+  // Extract just the filename without path and extension
+  const getName = (filepath: string) => {
+    const filename = filepath.split('/').pop() || filepath;
```
You also extract the file name at row 137
`  const file1Name = file1.split('/').pop()?.replace(/\.gpkg$/i, '') || 'file1';`

Therefore, this function should be scoped outside, and not as inner funtion. This way you would avoid code duplication. Make it fit all possible scenarios.

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  return `merged_${name1}_${name2}.gpkg`;
+}
+
+let output = createOutputFilename(file1, file2, args[2]);
```
`output` does not describe the meaning of the variable well enough.
make the name more meaningful.

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  return `merged_${name1}_${name2}.gpkg`;
+}
+
+let output = createOutputFilename(file1, file2, args[2]);
```
please explain why is it `let` (and do not write because variable is redeclared)

roicohen326 replied:
You're right , I was using let because I was reassigning the variable later for filename conflicts.
I can see now const will be a better fit , I've refactored it to use const .
Now the variable is set once and never changes.

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+
+let output = createOutputFilename(file1, file2, args[2]);
+
+// Reusable function to get data tables from any database
```
Same, we should avoid having notes in the code.
A good code should be self-explanatory without adding words and explanations.
(this is chatGPT's outcome... it's good using him we just need to modify him where needed)

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  `).all() as TableInfo[];
+}
+
+// Helper functions for table processing
```
Same, we should avoid having notes in the code.
A good code should be self-explanatory without adding words and explanations.
(this is chatGPT's outcome... it's good using him we just need to modify him where needed)

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  const createTableSQL = sourceDb.prepare(`
+    SELECT sql FROM sqlite_master 
+    WHERE type='table' AND name = ?
+  `).get(tableName) as any;
```
I don't like as any,
`prepere` output is typed `BetterSqlite3.Statement<unknown, unknown>`
Please configure it accordingly.

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  `).get(tableName) as any;
   
-  console.log(`📊 Syria has ${targetTables.length} data tables`);
+  if (createTableSQL?.sql) {
```
Once you do configure it you won't need `?` which is not best practice

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+    return true;
+  } else {
+    console.log(`Warning: Could not get table structure for ${tableName}, skipping...`);
+    return false;
```
According to the function name, the operation of creating a table (when it doesn't exist) failed when returning false. According to the "warning", it just couldn't get some structure. I didn't fully understand. Try making it more readable.

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+      return 0; // Skip if table creation failed
+    }
+    
+    // Insert data
```
no comment needed

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  console.log(`Processing ${file2Name} table: ${tableName}`);
+  
+  try {
+    // Create table if it doesn't exist
```
no comment needed

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  try {
+    // Create table if it doesn't exist
+    if (!createTableIfNotExists(sourceDb, targetDb, tableName)) {
+      return 0; // Skip if table creation failed
```
I believe this should be some sort of error instead of returning 0 with no logs

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+    
+  } catch (err) {
+    console.log(`Error with table ${tableName}: ${err}`);
+    return 0;
```
throw error do not return 0.
https://engineering.udacity.com/handling-errors-like-a-pro-in-typescript-d7a314ad4991

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  if (!fs.existsSync(file1)) {
+    const error = new Error(`${file1} file missing`);
+    (error as any).status = StatusCodes.BAD_REQUEST;
+    throw error;
```
nicer

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  if (!fs.existsSync(file2)) {
+    const error = new Error(`${file2} file missing`);
+    (error as any).status = StatusCodes.BAD_REQUEST;
+    throw error;
```
code duplication one right after the other my friend.
תתקן את זה זריז ואעשה כאילו לא ראיתי את זה

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  const size1 = (fs.statSync(file1).size / BYTES_TO_MB).toFixed(2);
+  const size2 = (fs.statSync(file2).size / BYTES_TO_MB).toFixed(2);
+  const file1Name = file1.split('/').pop()?.replace(/\.gpkg$/i, '') || 'file1';
+  const file2Name = file2.split('/').pop()?.replace(/\.gpkg$/i, '') || 'file2';
```
as mentioned

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  console.log(`Step 2: Copy base file from ${file1} to ${output}...`);
+  if (fs.existsSync(output)) {
+    const timestamp = Date.now();
+    const newOutput = output.replace('.gpkg', `_${timestamp}.gpkg`);
```
check out path.extname()

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  }
+  console.log(`Base file copied (${file1Name} as base)`);
+
+  console.log(`Step 3: Open databases (${file1Name} and ${file2Name})...`);
```
We don't need to log both before and after an action.
just log "opening databases...". (also adding `Step x` is driving me crazy. Let's say I added a step in the middle, I would need to increment each step value? big no no.... remember SOLID's O?)

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
+  const targetTables = getDataTables(targetDb);
+  const sourceTables = getDataTables(sourceDb);
+  
+  console.log(`${file1Name} has ${targetTables.length} data tables`);
```
Irrelevant info

razbroc commented on merge.ts:
```diff
@@ -1,116 +1,200 @@
   for (const table of sourceTables) {
-    const count = sourceDb.prepare(`SELECT COUNT(*) as count FROM ${(table as any).name}`).get() as any;
-    console.log(`   - ${(table as any).name}: ${count.count} rows`);
+    const count = sourceDb.prepare(`SELECT COUNT(*) as count FROM ${table.name}`).get() as CountResult;
```
`as number`? if CountResult eventually just holds a number...

razbroc replied:
code duplication for counting .

Also why are you counting just for logging?? logging should reflect what happens in the code. We do not write code for logging.

logs serves us, we do not serve the logs.

razbroc commented on package.json:
```diff
@@ -3,7 +3,8 @@
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "echo \"Error: no test specified\" && exit 1",
```
not yet....

razbroc reviewed (changes requested):
well done. mind the comments

roicohen326 committed:
docs: add usage instructions for CLI arguments in README

roicohen326 committed:
fix: eliminate code duplication and improve error handling
